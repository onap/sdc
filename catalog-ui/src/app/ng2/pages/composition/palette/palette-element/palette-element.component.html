<div class="palette-element" >
        <sdc-element-icon class="palette-element-icon" [iconName]="paletteElement.icon"
            [elementType]="paletteElement.componentSubType"[uncertified]="this.paletteElement.certifiedIconClass"></sdc-element-icon>
    <div class="palette-element-text">
        <div class="palette-element-name" sdc-tooltip
                            tooltip-text='{{paletteElement.name | resourceName}}'>{{paletteElement.name | resourceName}}
        </div>
        <span> V.{{paletteElement.version}}</span>
        <span>{{paletteElement.componentSubType}}</span>
    </div>
</div>