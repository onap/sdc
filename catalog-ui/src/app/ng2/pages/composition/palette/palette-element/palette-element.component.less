@import "./../../../../../../assets/styles/override";
.palette-element {
  cursor: pointer;
  display: flex;
  flex-direction: row;
  max-height: 65px;
  border-bottom: 1px solid @sdcui_color_silver;
  padding: 10px;
  align-items: center;
  .palette-element-icon {
    min-width: 45px;
    text-align: center;
  }

  .palette-element-text {
    display: flex;
    flex-direction: column;
    font-size: 13px;
    line-height: 15px;
    padding-left: 10px;
    font-family: OpenSans-Regular, sans-serif;
    overflow: hidden;

    .palette-element-name {
      color: @sdcui_color_dark-gray;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }
  }

}