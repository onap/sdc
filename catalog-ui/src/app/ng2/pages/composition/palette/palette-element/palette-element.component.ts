/**
 * Created by ob0695 on 6/28/2018.
 */
/*-
 * ============LICENSE_START=======================================================
 * SDC
 * ================================================================================
 * Copyright (C) 2017 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END=========================================================
 */

import {Component, Input} from "@angular/core";
import {LeftPaletteComponent} from "app/models/components/displayComponent";

@Component({
    selector: 'palette-element',
    templateUrl: './palette-element.component.html',
    styleUrls: ['./palette-element.component.less']
})
export class PaletteElementComponent {

    @Input() paletteElement: LeftPaletteComponent;
}
