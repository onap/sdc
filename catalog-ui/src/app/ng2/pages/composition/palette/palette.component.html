<div class="composition-palette-component">
    <div class="palette-elements-count">Elements
        <span class="palette-elements-count-value">{{numberOfElements}}</span>
    </div>

    <sdc-filter-bar placeholder="Search..." (valueChange)="onSearchChanged($event)" testId="searchAsset"></sdc-filter-bar>

    <div class="palette-elements-list">
        <sdc-loader [global]="false" name="palette-loader" testId="palette-loader" [active]="this.isPaletteLoading" [class.inactive]="!this.isPaletteLoading"></sdc-loader>
        <div *ngIf="numberOfElements === 0 && searchText" class="no-elements-found">No Elements Found</div>
        <sdc-accordion *ngFor="let mapByCategory of paletteElements | keyValue; let first = first" [attr.data-tests-id]="'leftPalette.category.'+mapByCategory.key" [title]="mapByCategory.key" [css-class]="'palette-category'">
            <div *ngFor="let mapBySubCategory of mapByCategory.value | keyValue">
                <div class="palette-subcategory">{{mapBySubCategory.key}}</div>
                <ng-container *ngIf="!(isViewOnly$ | async)">
                    <div *ngFor="let paletteElement of mapBySubCategory.value"
                        [dndDraggable]="paletteElement"
                        [dndDisableIf]="paletteElement.componentSubType == 'GROUP' && paletteElement.componentSubType == 'POLICY'"
                        (dndStart)="onDragStart($event, paletteElement)"
                        (drag)="onDraggableMoved($event)"
                        [dndEffectAllowed]="'copyMove'"
                        (mouseenter)="onMouseOver($event, paletteElement)"
                        (mouseleave)="onMouseOut(paletteElement)"
                        [attr.data-tests-id]="paletteElement.name">
                       <palette-element [paletteElement]="paletteElement"></palette-element>
                   </div>
                </ng-container>
                <ng-container *ngIf="(isViewOnly$ | async)">
                    <div *ngFor="let paletteElement of mapBySubCategory.value"
                        [attr.data-tests-id]="paletteElement.name">
                        <palette-element [paletteElement]="paletteElement"></palette-element>
                    </div>
                </ng-container>
            </div>
        </sdc-accordion>
    </div>
</div>

<div id="draggable_element" dndDropzone (dndDrop)="onDrop($event)" [dndAllowExternal]="true">
    <sdc-element-icon *ngIf="paletteDraggedElement" [iconName]="paletteDraggedElement.icon"
        [elementType]="paletteDraggedElement.componentSubType" [uncertified]="paletteDraggedElement.certifiedIconClass"></sdc-element-icon>
</div>
