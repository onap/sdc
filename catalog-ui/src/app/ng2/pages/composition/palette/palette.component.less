@import "./../../../../../assets/styles/override";

:host(composition-palette) {
  display:flex;
  flex: 0 0 244px;
}

sdc-loader.inactive {
  display:none;
}

:host ::ng-deep .sdc-filter-bar .sdc-input {
  margin-bottom:0px;
}
:host ::ng-deep .sdc-loader-wrapper {
  position:static; 
}

.composition-palette-component {
  background-color: @sdcui_color_white;
  overflow-y: auto;
  overflow-x: hidden;
  display: flex;
  flex-direction: column;
  position:relative;
  width: 244px;
  box-shadow: 7px -3px 6px -8px @sdcui_color_gray;

  .palette-elements-count {
    background-color: @sdcui_color_gray;
    line-height: 40px;
    padding: 0 17px;
    color: @sdcui_color_white;
    .palette-elements-count-value {
      float: right;
    }
  }

  .palette-elements-list {

    .no-elements-found {
      padding-left: 40px;
    }
    /deep/ .palette-category {
      display: flex;
      margin: 0px;
      .sdc-accordion-header {
        background-color: @sdcui_color_silver;
        margin: 0px;
        line-height: 40px;
        padding: 0px 10px;
      }
      .sdc-accordion-body {
        padding: 0px;
      }
    }
    .palette-subcategory {
      padding: 0 10px;
      background-color: @sdcui_color_lighter-silver;
      line-height: 35px;
    }
  }
}

#draggable_element {
  display: inline-block;
  border-radius: 50%;
  background: transparent;
  position: absolute;
  top: -9999px;
  left: 0;
  z-index: 100;
}

.invalid-drag {
  border: 7px solid @red-shadow;
}

.valid-drag {
  border: 7px solid @green-shadow;
}

@green-shadow: rgba(29, 154, 149, 0.3);
@red-shadow: rgba(218, 31, 61, 0.3);
