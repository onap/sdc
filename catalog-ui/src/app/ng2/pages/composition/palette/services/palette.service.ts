import { HttpClient, HttpParams } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { LeftPaletteComponent, LeftPaletteMetadataTypes } from 'app/models/components/displayComponent';
import { GroupMetadata } from 'app/models/group-metadata';
import { PolicyMetadata } from 'app/models/policy-metadata';
import { IComponentMetadata } from 'app/models/component-metadata';
import { SdcConfigToken } from 'app/ng2/config/sdc-config.config';
import { ISdcConfig } from 'app/ng2/config/sdc-config.config.factory';
import { WorkspaceService } from 'app/ng2/pages/workspace/workspace.service';
import 'rxjs/add/observable/forkJoin';
import { Observable } from 'rxjs/Rx';
import Dictionary = _.Dictionary;



@Injectable()
export class CompositionPaletteService {

    protected baseUrl = '';

    private leftPaletteComponents: Dictionary<Dictionary<LeftPaletteComponent[]>>;
    private facadeUrl: string;
    constructor(protected http: HttpClient, @Inject(SdcConfigToken) sdcConfig: ISdcConfig, private workspaceService: WorkspaceService) {
        this.baseUrl = sdcConfig.api.root + sdcConfig.api.component_api_root;
        this.facadeUrl = sdcConfig.api.uicache_root + sdcConfig.api.GET_uicache_left_palette;

    }

    public subscribeToLeftPaletteElements(next, error) {

        let params = new HttpParams();
        params = params.append('internalComponentType', this.workspaceService.getMetadataType());

        const loadInstances = this.http.get(this.facadeUrl, {params});
        const loadGroups = this.http.get(this.baseUrl + 'groupTypes', {params});
        const loadPolicies = this.http.get(this.baseUrl + 'policyTypes', {params});

        Observable.forkJoin(
            loadInstances, loadGroups, loadPolicies
        ).subscribe( ([resInstances, resGrouops, resPolicies]) => {
            const combinedDictionary = this.combineResoponses(resInstances, resGrouops, resPolicies);
            this.leftPaletteComponents = combinedDictionary;
            next(this.leftPaletteComponents);
        });
    }

    public getLeftPaletteElements = (): Dictionary<Dictionary<LeftPaletteComponent[]>> => {
        return this.leftPaletteComponents;
    }


    public convertPoliciesOrGroups = (paletteListResult, type: string ) => {
        const components: LeftPaletteComponent[] = [];

        if (type === 'Policies') {
            _.forEach(paletteListResult, (policyMetadata: PolicyMetadata) => {
                components.push(new LeftPaletteComponent(LeftPaletteMetadataTypes.Policy, policyMetadata));
            });
            return {
                Policies: components
            };
        }

        if (type === 'Groups') {
            _.forEach(paletteListResult, (groupMetadata: GroupMetadata) => {
                const item = new LeftPaletteComponent(LeftPaletteMetadataTypes.Group, groupMetadata);
                components.push(item);
            });
            return {
                Groups: components
            };
        }

        return {};
    }

    private combineResoponses(resInstances: object, resGrouops: object, resPolicies: object) {
        const retValObject = {};
        
        if (resInstances['Generic']) {
	        if (this.isSubstitutionForNestedServices()) {
	            const serviceGroup = this.createServiceGroup(resInstances);
	            if (serviceGroup) {
	                retValObject['Service'] = serviceGroup;
	            }
	        }
	        else {
	            // Generic will be the 1st category in the left Pallete
	        	retValObject['Generic'] = resInstances['Generic'];
	        }
	    }
        
        // Add all other categories
        for (const category in resInstances) {
            if (category === 'Generic') {
                continue;
            }
            retValObject[category] = resInstances[category];
        }

        // Add Groups
        retValObject["Groups"] = this.convertPoliciesOrGroups(resGrouops, 'Groups');

        // Add policies
        retValObject["Policies"] = this.convertPoliciesOrGroups(resPolicies, 'Policies');

        return retValObject;
    }
    
    private isSubstitutionForNestedServices(): boolean {
	    return this.workspaceService.metadata.categories[0].useServiceSubstitutionForNestedServices;
	}
    
    private createServiceGroup(resInstances: object): object {
	    const servicesList = resInstances['Generic']['Generic'];
	    if (Array.isArray(servicesList) && servicesList.length > 0) {
	        delete resInstances['Generic']['Generic'];
	        return servicesList.reduce(function (map, component) {
	            if (map[component.categories[0].name]) {
	                map[component.categories[0].name].push(component);
	            } else {
	                map[component.categories[0].name] = [component];
	            }
	            return map;
	        }, {});
	    }
	    return null;
	}
	
}
