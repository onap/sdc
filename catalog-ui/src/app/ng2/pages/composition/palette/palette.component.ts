/**
 * Created by ob0695 on 6/28/2018.
 */
/*-
 * ============LICENSE_START=======================================================
 * SDC
 * ================================================================================
 * Copyright (C) 2017 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END=========================================================
 */

import { Component, HostListener } from '@angular/core';
import { Select } from '@ngxs/store';
import { LeftPaletteComponent, LeftPaletteMetadataTypes } from 'app/models/components/displayComponent';
import { Point } from 'app/models/graph/point';
import { WorkspaceState } from 'app/ng2/store/states/workspace.state';
import Dictionary = _.Dictionary;
import { EventListenerService } from 'app/services/event-listener-service';
import { GRAPH_EVENTS } from 'app/utils/constants';
import { DndDropEvent } from 'ngx-drag-drop/ngx-drag-drop';
import { CompositionPaletteService } from './services/palette.service';
import {PolicyMetadata} from "../../../../models/policy-metadata";
import {GenericBrowserDomAdapter} from "@angular/platform-browser/src/browser/generic_browser_adapter";

@Component({
    selector: 'composition-palette',
    templateUrl: './palette.component.html',
    styleUrls: ['./palette.component.less']
})
export class PaletteComponent {

    constructor(private compositionPaletteService: CompositionPaletteService, private eventListenerService: EventListenerService) {}

    @Select(WorkspaceState.isViewOnly) isViewOnly$: boolean;
    private paletteElements: Dictionary<Dictionary<LeftPaletteComponent[]>>;
    public numberOfElements: number = 0;
    public isPaletteLoading: boolean;
    private paletteDraggedElement: LeftPaletteComponent;
    public position: Point = new Point();

    ngOnInit() {
        this.isPaletteLoading = true;

        this.compositionPaletteService.subscribeToLeftPaletteElements((leftPaletteElementsResponse) => {
            this.paletteElements = leftPaletteElementsResponse;
            this.numberOfElements = this.countLeftPalleteElements(this.paletteElements);
            this.isPaletteLoading = false;
        }, () => {
            this.isPaletteLoading = false;
        });

    }

    public buildPaletteByCategories = (searchText?: string) => { // create nested by category & subcategory, filtered by search parans
        // Flat the object and run on its leaves
        if (searchText) {
            searchText = searchText.toLowerCase();
            const paletteElementsAfterSearch = {};
            this.paletteElements = this.compositionPaletteService.getLeftPaletteElements();
            for (const category in this.paletteElements) {
                for (const subCategory in this.paletteElements[category]) {
                    const subCategoryToCheck = this.paletteElements[category][subCategory];
                    const res = subCategoryToCheck.filter((item) => item.searchFilterTerms.toLowerCase().indexOf(searchText) >= 0)
                    if (res.length > 0) {
                        paletteElementsAfterSearch[category] = {};
                        paletteElementsAfterSearch[category][subCategory] = res;
                    }
                }
            }
            this.paletteElements = paletteElementsAfterSearch;
        } else {
            this.paletteElements = this.compositionPaletteService.getLeftPaletteElements();
        }
        this.numberOfElements = this.countLeftPalleteElements(this.paletteElements);
    }

    public onSearchChanged = (searchText: string) => {

        if (this.compositionPaletteService.getLeftPaletteElements()) {
            this.buildPaletteByCategories(searchText);
        }
    }

    private countLeftPalleteElements(leftPalleteElements: Dictionary<Dictionary<LeftPaletteComponent[]>>) {
        // Use _ & flat map
        let counter = 0;
        for (const category in leftPalleteElements) {
            for (const subCategory in leftPalleteElements[category]) {
                counter += leftPalleteElements[category][subCategory].length;
            }
        }
        return counter;
    }

    private isGroupOrPolicy(component: LeftPaletteComponent): boolean {
        if (component &&
            (component.categoryType === LeftPaletteMetadataTypes.Group ||
            component.categoryType === LeftPaletteMetadataTypes.Policy)) {
            return true;
        }
        return false;
    }
    @HostListener('document:dragover', ['$event'])
    public onDrag(event) {
        this.position.x = event.clientX;
        this.position.y = event.clientY;
    }
    
    //---------------------------------------Palette Events-----------------------------------------//

    public onDraggableMoved = (event:DragEvent) => {  
          let draggedElement = document.getElementById("draggable_element");
          draggedElement.style.top = (this.position.y - 80) + "px";
          draggedElement.style.left = (this.position.x - 30) + "px";
          this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_DRAG_ACTION, this.position);
    }

    public onDragStart = (event, draggedElement:LeftPaletteComponent) => { // Applying the dragged svg component to the draggable element
        
        this.paletteDraggedElement = draggedElement;
        event.dataTransfer.dropEffect = "copy";
        let hiddenImg = document.createElement("span");
        event.dataTransfer.setDragImage(hiddenImg, 0, 0);
    }


    public onDrop = (event:DndDropEvent) => {
        let draggedElement = document.getElementById("draggable_element");
        draggedElement.style.top = "-9999px";
        if(draggedElement.classList.contains('valid-drag')) {
            if(!event.data){
                event.data = this.paletteDraggedElement;
            }
            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_DROP, event);
        } else {
            console.log("INVALID drop");
        }
        this.paletteDraggedElement = undefined;
        
    }

    public onMouseOver = (sectionElem:MouseEvent, displayComponent:LeftPaletteComponent) => {
        console.debug("On palette element MOUSE HOVER: ", displayComponent);
        if (this.isGroupOrPolicy(displayComponent)) {
            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_SHOW_POPUP_PANEL, displayComponent, sectionElem.target);
        } else {
            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HOVER_IN, displayComponent);
        }
    };

    public onMouseOut = (displayComponent:LeftPaletteComponent) => {
        console.debug("On palette element MOUSE OUT: ", displayComponent);
        if (this.isGroupOrPolicy(displayComponent)) {
            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HIDE_POPUP_PANEL);
        } else {
            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HOVER_OUT);
        }
    };

}
