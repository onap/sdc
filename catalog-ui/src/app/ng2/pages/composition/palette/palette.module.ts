import { NgModule } from "@angular/core";
import { CompositionPaletteService } from "./services/palette.service";
import { PaletteComponent } from "./palette.component";
import { SdcUiComponentsModule } from "onap-ui-angular";
import { GlobalPipesModule } from "app/ng2/pipes/global-pipes.module";
import { CommonModule } from "@angular/common";
import { DndModule } from "ngx-drag-drop";
import {PaletteElementComponent} from "./palette-element/palette-element.component";
import {EventListenerService} from "app/services/event-listener-service";
import {UiElementsModule} from "app/ng2/components/ui/ui-elements.module";

@NgModule({
    declarations: [PaletteComponent, PaletteElementComponent],
    imports: [CommonModule, SdcUiComponentsModule, GlobalPipesModule, UiElementsModule, DndModule],
    exports: [PaletteComponent],
    entryComponents: [PaletteComponent],
    providers: [CompositionPaletteService, EventListenerService]
})
export class PaletteModule {

    constructor() {

    }

}