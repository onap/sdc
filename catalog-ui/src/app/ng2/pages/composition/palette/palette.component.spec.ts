import {async, ComponentFixture, TestBed} from "@angular/core/testing";
import {NO_ERRORS_SCHEMA} from "@angular/core";
import {CompositionPaletteService} from "./services/palette.service";
import {EventListenerService} from "../../../../services/event-listener-service";
import {PaletteElementComponent} from "./palette-element/palette-element.component";
import {PaletteComponent} from "./palette.component";
import {ConfigureFn, configureTests} from "../../../../../jest/test-config.helper";
import {GRAPH_EVENTS} from "../../../../utils/constants";
import {KeyValuePipe} from "../../../pipes/key-value.pipe";
import {ResourceNamePipe} from "../../../pipes/resource-name.pipe";
import {LeftPaletteComponent} from "../../../../models/components/displayComponent";
import {Observable} from "rxjs/Observable";
import {leftPaletteElements} from "../../../../../jest/mocks/left-paeltte-elements.mock";
import {NgxsModule, Select} from '@ngxs/store';
import { WorkspaceState } from 'app/ng2/store/states/workspace.state';


describe('palette component', () => {

    const mockedEvent = <MouseEvent>{ target: {} }
    let fixture: ComponentFixture<PaletteComponent>;
    let eventServiceMock: Partial<EventListenerService>;
    let compositionPaletteMockService: Partial<CompositionPaletteService>;

    beforeEach(
        async(() => {
            eventServiceMock = {
                notifyObservers: jest.fn()
            }
            compositionPaletteMockService = {
                subscribeToLeftPaletteElements:  jest.fn().mockImplementation(()=>  Observable.of(leftPaletteElements)),
                getLeftPaletteElements: jest.fn().mockImplementation(()=>  leftPaletteElements)
            }
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [PaletteComponent, PaletteElementComponent, KeyValuePipe, ResourceNamePipe],
                    imports: [NgxsModule.forRoot([WorkspaceState])],
                    schemas: [NO_ERRORS_SCHEMA],
                    providers: [
                        {provide: CompositionPaletteService, useValue: compositionPaletteMockService},
                        {provide: EventListenerService, useValue: eventServiceMock}
                    ],
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(PaletteComponent);
            });
        })
    );

    it('should match current snapshot of palette component', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('should call on palette component hover in event', () => {
        let paletteObject =  <LeftPaletteComponent>{categoryType: 'COMPONENT'};
        fixture.componentInstance.onMouseOver(mockedEvent, paletteObject);
        expect(eventServiceMock.notifyObservers).toHaveBeenCalledWith(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HOVER_IN, paletteObject);
    });

    it('should call on palette component hover out event', () => {
        let paletteObject =  <LeftPaletteComponent>{categoryType: 'COMPONENT'};
        fixture.componentInstance.onMouseOut(paletteObject);
        expect(eventServiceMock.notifyObservers).toHaveBeenCalledWith(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HOVER_OUT);
    });

    it('should call show popup panel event', () => {
        let paletteObject =  <LeftPaletteComponent>{categoryType: 'GROUP'};
        fixture.componentInstance.onMouseOver(mockedEvent, paletteObject);
        expect(eventServiceMock.notifyObservers).toHaveBeenCalledWith(GRAPH_EVENTS.ON_PALETTE_COMPONENT_SHOW_POPUP_PANEL, paletteObject, mockedEvent.target);
    });

    it('should call  hide popup panel event', () => {
        let paletteObject =  <LeftPaletteComponent>{categoryType: 'GROUP'};
        fixture.componentInstance.onMouseOut(paletteObject);
        expect(eventServiceMock.notifyObservers).toHaveBeenCalledWith(GRAPH_EVENTS.ON_PALETTE_COMPONENT_HIDE_POPUP_PANEL);
    });

    it('should build Palette By Categories without searchText', () => {
        fixture.componentInstance.buildPaletteByCategories();
        expect(fixture.componentInstance.paletteElements["Generic"]["Network"].length).toBe(5);
        expect(fixture.componentInstance.paletteElements["Generic"]["Network"][0].searchFilterTerms).toBe("extvirtualmachineinterfacecp external port for virtual machine interface extvirtualmachineinterfacecp 3.0");
        expect(fixture.componentInstance.paletteElements["Generic"]["Network"][1].searchFilterTerms).toBe("newservice2 asdfasdfa newservice2 0.3");

        expect(fixture.componentInstance.paletteElements["Generic"]["Configuration"].length).toBe(1);
        expect(fixture.componentInstance.paletteElements["Generic"]["Configuration"][0].systemName).toBe("Extvirtualmachineinterfacecp");
    });

    it('should build Palette By Categories with searchText', () => {
        fixture.componentInstance.buildPaletteByCategories("testVal");
        expect(fixture.componentInstance.paletteElements["Generic"]["Network"].length).toBe(1);
        expect(fixture.componentInstance.paletteElements["Generic"]["Network"][0].searchFilterTerms).toBe("testVal and other values");
    });

    it('should change numbers of elements', () => {
        fixture.componentInstance.buildPaletteByCategories();
        expect(fixture.componentInstance.numberOfElements).toEqual(6);
        fixture.componentInstance.buildPaletteByCategories("testVal");
        expect(fixture.componentInstance.numberOfElements).toEqual(1);
    });
});