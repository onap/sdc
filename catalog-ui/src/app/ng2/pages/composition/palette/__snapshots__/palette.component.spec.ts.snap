// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`palette component should match current snapshot of palette component 1`] = `
<composition-palette
  buildPaletteByCategories={[Function Function]}
  compositionPaletteService={[Function Object]}
  eventListenerService={[Function Object]}
  numberOfElements="0"
  onDragStart={[Function Function]}
  onDraggableMoved={[Function Function]}
  onDrop={[Function Function]}
  onMouseOut={[Function Function]}
  onMouseOver={[Function Function]}
  onSearchChanged={[Function Function]}
  position={[Function Point]}
>
  <div
    class="composition-palette-component"
  >
    <div
      class="palette-elements-count"
    >
      Elements 
      <span
        class="palette-elements-count-value"
      >
        
      </span>
    </div>
    <sdc-filter-bar
      placeholder="Search..."
      testid="searchAsset"
    />
    <div
      class="palette-elements-list"
    >
      <sdc-loader
        name="palette-loader"
        testid="palette-loader"
      />
      
      
    </div>
  </div><div
    dnddropzone=""
    id="draggable_element"
  >
    
  </div>
</composition-palette>
`;
